### PR TITLE
Add use of libabigail for ABI checks

### PR DESCRIPTION
**Pull request by bregma**
Added `make -C src check-abi` target to leverage libabigail 2.0 tools to check for ABI changes.

Add ABI baseline files for aarch64, i686, riscv, s390x, and x86_64 Linux targets and aarch64 and x86_64 QNX SDP 7.1 targets.

Used the check-abi target in the CI-unix github workflow.

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -1248,10 +1248,44 @@ AM_CPPFLAGS = -I$(top_srcdir)/include -I$(top_srcdir)/include/tdep-$(arch) -I.
...
+               --drop-private-types
+
+abi-generate: $(lib_LTLIBRARIES)
+	@printf 'Generating ABI files\n'; \
+	mkdir -p abi; \
+	for lib in $(lib_LTLIBRARIES:.la=); do \
+		sofile="$$(find . -name $${lib}.so)"; \
+		abifile="abi/$${lib}-${arch}-${target_os}.abi"; \
+		abidw "$${sofile}" $(abidw_common) --out-file "$${abifile}"; \
+	done
```
Comment (bertwesarg) [inline src/Makefile.am:1276]: honor silient rules:

```suggestion
	@mkdir -p abi; \
	for lib in $(lib_LTLIBRARIES:.la=); do \
		sofile="$$(find . -name $${lib}.so)"; \
		abifile="abi/$${lib}-${arch}-${target_os}.abi"; \
		if $(AM_V_P); then echo abidw "$${sofile}" $(abidw_common) --out-file "$${abifile}"; \
		              else echo "  GEN ABI  $${abifile}"; \
		fi; \
		abidw "$${sofile}" $(abidw_common) --out-file "$${abifile}"; \
	done
```

**Comment by bregma** [comment]
Found a better, more elegant way to generate and process the ABI files by using **make** rules instead of shell scripting.

File: src/Makefile.am (hunk 1 above)
Comment (bregma) [reply src/Makefile.am]: See latest revision.

**Comment by bertwesarg** [comment]
> Found a better, more elegant way to generate and process the ABI files by using **make** rules instead of shell scripting.

nice. just in case you want a custom prefix instead of the generic `GEN`. put something like this into your Makefile.am:

```
UNW_V_ABIDW = $(unw__v_ABIDW_@AM_V@)
unw__v_ABIDW_ = $(unw__v_ABIDW_@AM_DEFAULT_V@)
unw__v_ABIDW_0 = @echo "  ABIDW    $@";
unw__v_ABIDW_1 =
UNW_V_ABIDIFF = $(unw__v_ABIDIFF_@AM_V@)
unw__v_ABIDIFF_ = $(unw__v_ABIDIFF_@AM_DEFAULT_V@)
unw__v_ABIDIFF_0 = @echo "  ABIDIFF  $@";
unw__v_ABIDIFF_1 =
```

and use `$(UNW_V_ABIDW)` and `$(UNW_V_ABIDIFF)`

File: src/Makefile.am [hunk 2]
```diff
@@ -1248,10 +1248,39 @@ AM_CPPFLAGS = -I$(top_srcdir)/include -I$(top_srcdir)/include/tdep-$(arch) -I.
...
 MAINTAINERCLEANFILES = Makefile.in
 
+#
+# ABI checks
+#
+# Running the ABI checks require configuring with CFLAGS="-g -Og" and running
+# the targets explicitly with `make -C src abi-check`. The result is output in a
+# file with the `.abidiff` extension.
+#
+# This also requires libabigail tools of at least version 2.0 to be installed.
```
Comment (bertwesarg) [inline src/Makefile.am:1262]: Ubunt 20.04 TLS  still includes 1.6. I think that justifies a configure check, which then also includes the `--enable-shared` check

File: src/Makefile.am [hunk 3]
```diff
@@ -1248,10 +1248,39 @@ AM_CPPFLAGS = -I$(top_srcdir)/include -I$(top_srcdir)/include/tdep-$(arch) -I.
...
+#
+.la.abi:
+	$(AM_V_GEN) soname=$$(sed -n -e "s#dlname='\(.*\)\.[0-9]*'#\1#p" $<); \
+	if test -n "$$soname"; then \
```
Comment (bertwesarg) [inline src/Makefile.am:1266]: test for the existens of `test -f .libs/$${soname}`?

**Comment by bregma** [comment]
The archived ABI files need to be generated against a build configured with `--enable-debug` because that configuration flag **changes the ABI** and it's used in CI. Maybe there should be two sets of archived ABI files, one for with `--enable-debug` and one for without?

File: src/Makefile.am [hunk 4]
```diff
@@ -1248,10 +1248,39 @@ AM_CPPFLAGS = -I$(top_srcdir)/include -I$(top_srcdir)/include/tdep-$(arch) -I.
 AM_CCASFLAGS = $(AM_CPPFLAGS)
 noinst_HEADERS += unwind/unwind-internal.h
 
-EXTRA_DIST = $(libunwind_la_EXTRAS_ia64)
+EXTRA_DIST = $(libunwind_la_EXTRAS_ia64) abi
 
 MAINTAINERCLEANFILES = Makefile.in
 
+#
```
[the remark this answers is not in the document]
Comment (bregma) [reply src/Makefile.am:1258]: Now it'll just work (and do nothing) with `--disable-shared`.

File: src/Makefile.am (hunk 3 above)
Comment (bregma) [reply src/Makefile.am:1269]: Done.

File: src/Makefile.am (hunk 2 above)
Comment (bregma) [reply src/Makefile.am:1265]: Done.

**Comment by bertwesarg** [comment]
LGTM

**Pull request merged**